kong/conf_loader.lua

```diff
@@ -23,6 +23,7 @@ local DEFAULT_PATHS = {
   "/etc/kong.conf",
 }
 
+local secret_path = "/etc/kong/secrets"
```

Comment:
would it make sense to make this value configurable?

Reply:
It would, though looking at that I found it resulted in a "chicken and egg" problem. 

Since the override with file secrets is done inside the confloader object, a configurable file secret path would itself be part of the configuration file.  So while loading the configuration the configuration is referenced.  

I couldn't find any instances of the configuration referencing itself yet, so I figured a constant file path would be a good first step.

Any suggestions on this would be welcome.

Reply:
Hi @jdevalk2 

Sorry to comment this Pull Request as a pure potential user of this feature, as I’m not an official reviewer!

I was about to comment that having a hard coded path for secret files is not that a good idea as security constraints on deployment, or rules for hardening Docker images (in case Kong is deployed on Docker) might mandate such configuration information to be located in specific directories other than `/etc/kong/secrets/`.

However, digging into your code, I have a second comment (first one remains applicable :P ): your proposed implementation looks to me not homogeneous with the configuration management already in place in Kong. While configuration is currently based on a file containing `tag = value` pairs, your proposal is to get sensitive data in a different manner: value of _each_ configuration parameter would be stored in an _individual_ file named `/etc/kong/secrets/<parameter value>`. 10 sensitive parameters => 10 files (as far as I understood; sorry if my understanding is incorrect).

The feature of providing (sensitive) configuration data as Kubernetes secrets definitely makes sense, but why not addressing it simply by the capability to get several configuration files as input to Kong (a kind of `-c <file1> -c <file2>`)? That would be a deployment decision to put some configuration data in clear text in `<file1>` (and in directory A), and the sensitive configuration data in a "protected" file `<file2>`, made available for example as a K8S secret file mounted as a volume (in directory B).
Priority would be linked to the precedence of the file within the command line (`<file1>` is in first position in the command line, so with higher priority in case data are part of both files… or the reverse ;) ). Usage of envvars would anyway be with the highest priority (a difference with your code).

This looks more in line with current configuration mechanism, and less linked to a pure Kubernetes use case, by the way.

Reply:
I think one distinction to make in the discussion is between Configuration and Secrets.
Secrets are a type of Configuration. But a different class of information.

Secrets like username / password have a different lifecycle than normal configuration.
For example I can put my username + password on rotation every week, or I can revoke a password.

Further as stated I am a personal believer that secrets should not be passed into environment variables.
It will not "break the bank" but I am a believer of Defense in Depth strategies and for purpose of environment dumps and that its so easy for them to end up in log files, they should not be in there.


To best of my knowledge (as I am not a Kubernetes guru)
In scope of Kubernetes, configmaps are used for configuration and non-secure data.
Kubernetes secrets are used for secure bits of data. Secrets are stored as a different class of data in Kubernetes (and also Docker)
And third party products exist to manage secrets and plug into kubernetes like Hashicorp Vault.


If we provide a configfile to Kong on the command line
Kong.exe -c baseconf.conf -c secretconf.conf


The control of which config overrides another would be part of the command line. 
The secretconf.conf would have to come from somewhere. 
Two ways could be it would have to be mounted as a file by a configmap, or it would have to be created by an init container.

As the configmap is not really meant for secret data, I feel the better would be init container.

In which case one could mount the secrets in the init container and have it produce that file.

I think I just argued against my own pull request!
Anyway from a docker / kubernetes perspective it would be nice to have an alternate other than having secrets passed into ENV variables.   I will explore the init container route a bit further as well.